File: demos_code_J1_conception_objet/02_ocp_strategie/DemoOcp.java
public class DemoOcp {
    record Commande(int totalCents) {}

    interface CalculFraisLivraison {
        int calculer(Commande commande);
    }

    static final class Standard implements CalculFraisLivraison {
        public int calculer(Commande c) {
            return c.totalCents() >= 5000 ? 0 : 500;
        }
    }

    static final class Express implements CalculFraisLivraison {
        public int calculer(Commande c) {
            return 1200;
        }
    }

    static final class Relais implements CalculFraisLivraison {
        public int calculer(Commande c) {
            return 300;
        }
    }

    // Extension ajoutée sans modifier totalAvecLivraison().
    static final class Drone implements CalculFraisLivraison {
        public int calculer(Commande c) {
            return 1800;
        }
    }

    static int totalAvecLivraison(Commande c, CalculFraisLivraison calcul) {
        return c.totalCents() + calcul.calculer(c);
    }

    public static void main(String[] args) {
        Commande c = new Commande(4200);

        System.out.println("Standard -> " + totalAvecLivraison(c, new Standard()));
        System.out.println("Express  -> " + totalAvecLivraison(c, new Express()));
        System.out.println("Relais   -> " + totalAvecLivraison(c, new Relais()));
        System.out.println("Drone    -> " + totalAvecLivraison(c, new Drone()));
    }
}

File: Demo en Python/demos_code_J1_conception_objet_python/02_ocp_strategie/DemoOcp.py
from abc import ABC, abstractmethod
from dataclasses import dataclass


@dataclass(frozen=True)
class Commande:
    total_cents: int


class CalculFraisLivraison(ABC):
    @abstractmethod
    def calculer(self, commande: Commande) -> int:
        pass


class Standard(CalculFraisLivraison):
    def calculer(self, commande: Commande) -> int:
        return 0 if commande.total_cents >= 5000 else 500


class Express(CalculFraisLivraison):
    def calculer(self, commande: Commande) -> int:
        return 1200


class Relais(CalculFraisLivraison):
    def calculer(self, commande: Commande) -> int:
        return 300


# Extension ajoutée sans modifier total_avec_livraison().
class Drone(CalculFraisLivraison):
    def calculer(self, commande: Commande) -> int:
        return 1800


def total_avec_livraison(c: Commande, calcul: CalculFraisLivraison) -> int:
    return c.total_cents + calcul.calculer(c)


def main() -> None:
    c = Commande(4200)

    print(f"Standard -> {total_avec_livraison(c, Standard())}")
    print(f"Express  -> {total_avec_livraison(c, Express())}")
    print(f"Relais   -> {total_avec_livraison(c, Relais())}")
    print(f"Drone    -> {total_avec_livraison(c, Drone())}")


if __name__ == "__main__":
    main()
